### ISSUE TITLE
Composer json parsing doesn't account for scripts described as objects

### ISSUE THREAD

**Issue by cedricziel** [heroku/buildpacks-php#81]
Usually, composer scripts are arrays or sequences of commands. In symfony projects, a set of framework provided scripts is defined as objects:

```json
{
    "scripts": {
        "auto-scripts": {
            "cache:clear": "symfony-cmd",
            "assets:install %PUBLIC_DIR%": "symfony-cmd",
            "importmap:install": "symfony-cmd"
        },
        "post-install-cmd": [
            "@auto-scripts"
        ],
        "post-update-cmd": [
            "@auto-scripts"
        ]
    }
}
```

This buildpack fails to build when a script is described as object.

### Proposed solution

Allow scripts to be defined as objects as well

**Comment by dzuelke** [comment]
Hi @cedricziel!

Do you have further examples of this usage pattern? What exactly does the key in the object achieve? AFAICT it's simply ignored, Composer will only run the value part with the command string?

**Comment by cedricziel** [comment]
Hi @dzuelke 

The CNB fails to parse the composer.json file entirely and can't complete the build. I think it's less an issue of with composer and more an issue with the way this build pack gathers metadata.

**Comment by dzuelke** [comment]
Yeah, I understand what's happening, @cedricziel.

My question was what this object notation is:

```json
        "auto-scripts": {
            "cache:clear": "symfony-cmd",
            "assets:install %PUBLIC_DIR%": "symfony-cmd",
            "importmap:install": "symfony-cmd"
        },
```

Every value in the `scripts` object in `composer.json` should be either a string, or an array of strings, not an object. It works because Composer loads the JSON in array mode, so it can iterate over the values.

But Composer does not use the keys you provide there; this object syntax is not documented in the Composer docs, and the keys are not used inside Composer's `EventDispatcher::doDispatch()`.

I researched, and it's a Symfony Flex thing, it magically expands these entries: https://github.com/symfony/flex/blob/4dc11919791f81d087a12db2ab4c7e044431ef6b/src/ScriptExecutor.php#L79-L91

I'll see what I can do to support this.

**Comment by cedricziel** [comment]
That would be great - I imagine this is a potential blocker for a large share of modern PHP apps and resolving it could help others adopt the CNB

**Comment by dzuelke** [comment]
> That would be great - I imagine this is a potential blocker for a large share of modern PHP apps and resolving it could help others adopt the CNB

Yup for sure. Just not the easiest of things given how we're using `serde` to parse `composer.json` to Rust types, and Rust is quite particular about its types :)

**Comment by cedricziel** [comment]
I looked at it before I opened the ticket, but had just started with rust. 

I think a struct for scripts with a custom serializer / deserializer could do that.

**Comment by schneems** [comment]
Related internal ticket number 1543786

> Yup for sure. Just not the easiest of things given how we're using serde to parse composer.json to Rust types, and Rust is quite particular about its types :)

You can parse into a `serde_json::Value` which is basically like dumping it into a Ruby hash. Playground link https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=d91ee7cc59b18dc4b626f2b4f7f37103. 

<details><summary>code for backup reference</summary>
<pre>
$ cat src/main.rs
fn main() {
    let data = r#"
{
    "scripts": {
        "auto-scripts": {
            "cache:clear": "symfony-cmd",
            "assets:install %PUBLIC_DIR%": "symfony-cmd",
            "importmap:install": "symfony-cmd"
        },
        "post-install-cmd": [
            "@auto-scripts"
        ],
        "post-update-cmd": [
            "@auto-scripts"
        ]
    }
}
    "#;
    let value: serde_json::Value = serde_json::from_str(data).unwrap();
    let content = value
        .get("scripts")
        .and_then(|value| value.get("auto-scripts"))
        .and_then(|value| value.as_object())
        .map(|object| {
            //
            object
                .keys()
                .map(|key| key.to_string())
                .collect::<Vec<String>>()
        });

    println!("{:?}", content);
}
$ cargo run
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.00s
     Running `target/debug/demo`
Some(["assets:install %PUBLIC_DIR%", "cache:clear", "importmap:install"])
</pre>
</details>

**Comment by schneems** [comment]
I believe #168, along with stacked PR #169, is the best path forward, but it's not the only path forward. Here are the existing proposed solutions:

## PR #132

Sent 209 days ago, this change is +3,065/−4 which seems huge, but when you remove test fixtures, tests, and the changelog the actual logic is ~30 lines of code:

```rust
impl ComposerBasePackage {
    #[must_use]
    pub fn get_script(&self, key: &str) -> Option<&ScriptValue> {
        self.scripts.as_ref().and_then(|scripts| scripts.get(key))
    }

    #[must_use]
    pub fn scripts_len(&self) -> usize {
        self.scripts.as_ref().map_or(0, HashMap::len)
    }

    #[must_use]
    pub fn has_script(&self, key: &str) -> bool {
        self.scripts.as_ref().map_or(false, |scripts| scripts.contains_key(key))
    }
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[serde(untagged)]
pub enum ScriptValue {
    String(String),
    Vec(Vec<String>),
    ScriptObject(HashMap<String, String>),
}
```

```rust
struct ComposerBasePackage {
    // ...
    pub scripts: Option<HashMap<String, ScriptValue>>,
}
```

## PR #168

I sent this yesterday. It's +109/-16. It includes an integration test case and some additional quality of life and fixes (like doc updates) I encountered along the way split out into different commits. You can also make this PR "net" smaller by accepting #169 (+4/-28), which uses the same fix but removes additional unused code.

## PR #170

Just opened this today, it's the same fix as #168 but without any extra commits. It is +0/-2. Removing these two lines fixes the problem:

```rust
    #[serde_as(as = "Option<HashMap<_, OneOrMany<_, PreferOne>>>")]
    pub scripts: Option<HashMap<String, Vec<String>>>,
```

## Some future PR

I believe that both #168 and #170 contain elegant solutions to this problem. I'll also hold space that some other solution could solve this in some other way, or some combination of the above approaches. Such as #170 and #169 (though I think we should add some test for this).

Unless there's a strong reason to parse "scripts" today, I feel we should remove it. If there are planned features that rely on scripts, we should adopt something like #132 when that information is needed. If that path is taken in the future and we want to add a symphony integration test, then we could either put it in an external repository and download it at CI setup time (similar to `hatchet setup` in the classic buildpack) or we could install the Symfony CLI and generate a new project like `symfony new my_project_directory --version="7.2.x" --webapp` https://symfony.com/doc/current/setup.html#creating-symfony-applications. This approach would help reduce the amount of checked-in code.

**Comment by schneems** [comment]
This is now released with `heroku/php` 0.2.3
